### PR TITLE
Uniswap v2 ink! smart contract

### PR DESCRIPTION
**Pull request by frankli-dev**
One thing to clarify here is that the max value of `FixedU128<U32>` is `79_228_162_514_264_337_593_543_950_335.9999999998`.

So for 18 decimals tokens, it can represent to `7.9 * 10^11`

### CODE REVIEW

File: contracts/uniswap_v2/lib.rs [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+        pub fn get_dex_incentive(
+            &self,
+            currency_id_a: TokenAddress,
+            currency_id_b: TokenAddress,
+            account: AccountId,
+        ) -> Balance {
+            if let Some(trading_pair) = TradingPair::from_currency_ids(currency_id_a, currency_id_b)
+            {
+                match self.dex_incentives.get(&(trading_pair, account)) {
+                    Option::Some(p) => *p,
```
Comment (simonsso) [inline contracts/uniswap_v2/lib.rs:782]: I prefer `Some()` over `Option::Some()`

File: contracts/uniswap_v2/lib.rs [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+        use crate::{
+            constants::{GET_EXCHANGE_FEE, TRADING_PATH_LIMIT},
+            models::{ExchangeRate, Ratio, TokenAddress, TradingPair},
+        };
+        use primitive_types::H160;
+
+        /// We test a simple use case of our contract.
+        #[test]
+        fn add_liquidity_works() {
+            // struct MockedExtension;
```
Comment (simonsso) [inline contracts/uniswap_v2/lib.rs:858]: This test case is totally useless.

**Review by simonsso — CHANGES_REQUESTED** [review]
Please increase test coverage, a test case with all code commented out is not adding any value to anyone.

File: contracts/uniswap_v2/lib.rs [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
+
+            self.env()
+                .extension()
+                .deposit(trading_pair.first(), caller, pool_0_increment)?;
+            self.env()
+                .extension()
+                .deposit(trading_pair.second(), caller, pool_1_increment)?;
+
+            self.do_deposit_pool(&trading_pair, pool_0_increment, pool_1_increment)?;
+            self.do_deposit_dex_share(caller, &trading_pair, share_increment)?;
```
[the remark this answers is not in the document]
Comment (frankli-dev) [reply contracts/uniswap_v2/lib.rs:661]: tested!

**Review by simonsso — APPROVED** [review]

**Pull request merged**